_posts/2019-05-06-live-view-with-presence.md

```diff
@@ -0,0 +1,587 @@
+
+Lastly, we need to teach our live view how to respond to this broadcast with a `handle_info/2` function:
+
+```elixir
+# lib/phat_web/live/chat_live_view.ex
+
+def handle_info(%{event: "message", payload: state}, socket) do
+  {:noreply, assign(socket, state)}
+end
+```
+
+The live view handles the `"message"` message by updating the socket's state with `%{chat: chat}` payload, where the chat is our newly updated chat containing the new message from the user. And that is all it takes to ensure that all subscribing clients see any new messages submitted into the chat template's new message form!
+
+## Tracking Users with Phoenix Presence
+
+Now that our live view is smart enough to broadcast messages to all of the users in the given chat room, we're ready to build some features that track and interact with those users. Let's say we want to have our template render a list of users in the chat room, something like this:
+
+![]({% asset live-view-presence-1.png @path %})
+
+We could create our own data structure for tracking user presence in a live view, store it in the live view's socket, and hand-roll our own functions to update that data structure when a user joins, leaves or otherwise changes their state. However, the [Phoenix Presence behaviour](https://hexdocs.pm/phoenix/Phoenix.Presence.html) abstracts this work away from us. It provides presence tracking for processes and channels, leveraging Phoenix PubSub behind the scenes to broadcast updates. It also uses a CRDT (Conflict-free Replicated Data Type) model, which means it works on distributed applications.
+
+Now that we understand a bit about what Presence is and why we want to use it, let's get it set up in our application.
+
+### Setting Up Presence
+
+In order to leverage Presence in our Phoenix app, we need to define our very our module:
+
+```elixir
+# lib/phat_web/presence.ex
+
+efmodule PhatWeb.Presence do
+  use Phoenix.Presence,
+    otp_app: :phat,
+    pubsub_server: Phat.PubSub
+end
+```
+
+The `PhatWeb.Presence` module does three things:
+
+* `uses` the Presence behaviour
+* Specifies that it shares a PubSub server with the rest of the application
+* Specifies that is shares our app's OTP app, which holds our application configuration
+
+Now we can use the `PhatWeb.Presence` module throughout our app to track user presence in a given process.
+
+### Tracking User Presence
+
+Our Presence module will maintain lists of present users in a given chat room by storing these users under a given topic of `"chat:#{chat_id}"`.
+
+So, when should we tell Presence to start tracking a given user? Well, at what point in time do we consider a user to be "present" in a chat room? When the user mounts the live view!
+
+We'll hook into our `mount/2` function to add the new user to Presence's list of users in a given chat room:
+
+```elixir
+# lib/phat_web/live/chat_live_view.ex
+
+defp topic(chat_id), do: "chat:#{chat_id}"
```

Comment:
Since this doesn't change from the last time we saw it, you could probably ommit it.